tests/integration/internal/test_lock.py

```diff
@@ -0,0 +1,69 @@
+import random
+import threading
+
+from spccore.internal.lock import *
+
+
+def test_private_acquire():
+    user1_lock = Lock("foo", max_age=datetime.timedelta(seconds=2))
+    user2_lock = Lock("foo", max_age=datetime.timedelta(seconds=2))
+    assert user1_lock._has_lock() is False
+    assert user2_lock._has_lock() is False
+
+    assert user1_lock._acquire_lock()
+    assert user1_lock._get_age() < 2
+    assert user2_lock._acquire_lock() is False
+
+    user1_lock.release()
+
+    assert user2_lock._acquire_lock()
+    assert user1_lock._acquire_lock() is False
+
+    user2_lock.release()
+
+
+def test_context_manager():
+    user1_lock = Lock("foo", max_age=datetime.timedelta(seconds=2))
+    user2_lock = Lock("foo", max_age=datetime.timedelta(seconds=2))
+
+    with user1_lock:
+        assert user1_lock._has_lock()
+        assert user1_lock._get_age() < 2
+        assert user2_lock._acquire_lock() is False
+
+    assert user1_lock._has_lock() is False
+    assert user2_lock._has_lock() is False
+
+
+def test_lock_timeout():
+    user1_lock = Lock("foo", max_age=datetime.timedelta(seconds=1))
+    user2_lock = Lock("foo", max_age=datetime.timedelta(seconds=1))
+
+    with user1_lock:
+        assert user1_lock._has_lock()
+        assert user1_lock._get_age() < 1.0
+        assert user2_lock._acquire_lock(break_old_locks=True) is False
+        time.sleep(2)
+        assert user1_lock._get_age() > 1.0
+        assert user2_lock._acquire_lock(break_old_locks=True)
+
+    assert user2_lock._has_lock()
+
+
+def test_renew():
+    user_lock = Lock("foo", max_age=datetime.timedelta(seconds=2))
+
+    with user_lock:
+        time.sleep(1.1)
+        assert user_lock._get_age() > 1.0
+        user_lock.renew()
+        assert user_lock._get_age() < 1.0
+
+
+def test_renew_with_expired():
+    user_lock = Lock("foo", max_age=datetime.timedelta(seconds=1))
+    with user_lock:
+        time.sleep(1.1)
+        assert user_lock._get_age() > 1.0
+        # expired but has not released
+        assert user_lock.renew() is False
```

Comment:
@zimingd , upon discussion with Bruce, the `test_multi_threaded` actually doesn't add any values. I removed `with lock` to test multiple threads doing the same task without acquiring the same lock, and get the same result with acquiring the same lock. I tested with 16 threads, each have done 100 tasks, and the result is the same. This confirms that the test passes regardless of weather the threads are using the same lock. Bruce recommended to remove that test since it doesn't add any value.
In general, we feel that the existing integration test cover more detailed cases where each thread may intersect at different section in the lock implementation.
Please do let me know if you recommend a different multi-thread test that fails if all threads do not attempt to acquire the same lock.